### ISSUE TITLE
New Feature: Rounded Corners

### ISSUE THREAD

**Issue by Javyre** [baskerville/bspwm#848]
I'd love to have rounded corners for non-maximized windows (for purely aesthetic reasons of course).
I have a WIP fork for this and am wondering if you'd be interested in a PR for this.

This would require to compile with the libxcb-shape extension library but we can check for compatibility for the protocol extension at runtime.

Relevant links:
- [shape protocol spec](https://www.x.org/releases/current/doc/xextproto/shape.html)
- [xlib docs could be useful](https://www.x.org/releases/X11R7.6/doc/libXext/shapelib.html)
- [a similar patch for openbox using xlib](https://forums.bunsenlabs.org/viewtopic.php?pid=46609#p46609)

**Comment by Javyre** [comment]
The X shape extension protocol allows only masking the shape with 1bit deep pixmaps.
So this results in jagged "rounded corners"
This is not *too* noticeable with a high resolution display but I'm looking deeper into the issue...
![](https://i.imgur.com/83Ufhjs.png)

[See my entry on the xcb mailing list](https://lists.freedesktop.org/archives/xcb/2018-September/011140.html)

**Comment by Matrix89** [comment]
Does this work with borders?

**Comment by Javyre** [comment]
![of course](https://i.imgur.com/tZIO5ex.png)

**Comment by Matrix89** [comment]
Im sold. I was trying to implement something like this a year ago but didn't have time to learn xlib and the rest of things. I've also implemented dobule borders. Unfortunately my implementation was really glithy.

**Comment by Javyre** [comment]
I'm interested in seeing the code for that.
I'm guessing its something along the lines of having a custom pixmaps set for the background layer

**Comment by Matrix89** [comment]
I can't find the code, sorry. I must have lost it during a reinstall

**Comment by Matrix89** [comment]
Can you share your code?

**Comment by Javyre** [comment]
The code is over on [my fork](https://github.com/Javyre/bspwm/)

Check out the commit history and you'll see the changes.
The first commit is a fake-fullscreen feature that might eventually deserve its own PR but all the other commits are relevant to this issue.

**Comment by Matrix89** [comment]
Im getting a seg fault [here](https://github.com/Javyre/bspwm/blame/master/src/events.c#L152) when launching firefox

**Comment by Javyre** [comment]
should be fixed with https://github.com/Javyre/bspwm/commit/faa4bfe928339318c9b0a504492699e138ae72be

**Comment by Javyre** [comment]
If anybody would like to know what the status is on the antialiasing issue I mentioned above see @psychon's response https://github.com/Javyre/bspwm/commit/94ba788802ea1d5d0e4419f7b9a357574b1e1451#r30542158

Basically, it wouldn't be possible without turning bspwm into a reparenting wm. And it would then require a compositor to be running.

Although I'm willing to implement this, it seems like a very intrusive change to bspwm and would like to hear from @baskerville before starting anything crazy.

**Comment by baskerville** [comment]
> Basically, it wouldn't be possible without turning bspwm into a reparenting wm.

I'm fine with the idea of having anti-aliased rounded corners.

The preselection feedback will also need to become partially rounded.

**Comment by Javyre** [comment]
Great then!
How do you see the antialiased corner option being implemented?

The way to do the antialiased corners is pretty different from the way my current code does it (i.e.: with a combination of the RENDER and SHAPE extension)

I'm thinking maybe keep the current code for users using no compositor and have a separate PR for the antialiased implementation. This would then require having `border_radius` (int) and `border_antialias` (bool)...

I will open a PR for some more fine grained code review discussion once I fix preselection feedback areas.

**Comment by Javyre** [comment]
Preselection feedback is now properly rounded https://github.com/Javyre/bspwm/commit/d925c6af0fb3d2acb1f5433a9f2649e3a96af136

**Comment by Javyre** [comment]
Non-gapless monocle is a weird case... I'm not sure whether it should or shouldn't have rounded corners since it has gaps around the window but border isn't drawn...

**Comment by psychon** [comment]
There is `xcb_shape_select_input` that allows you to ask the X11 server to send you `ShapeNotify` events when the shape of a window changes. Since you will also get this event for your own requests, you can have a counter per window (initialized to zero) that you increase each time you modify the shape and decrease each time you get a `ShapeNotify` event. If the counter becomes negative, "something else" is also changing the shape.
However, this will only detect the situation after "the damage" is already done...

I just checked with `xtrace` / `x11trace`: `xeyes` sets a shape before mapping its window. Actually, it has to, to avoid some flickering. So, by the time you get a `MapRequest`, you can check via `xcb_shape_query_extents` if the window already has a shape and then not set any shape (answer has fields `bounding_shaped` and `clip_shaped`).

**Comment by Javyre** [comment]
The second option seems less hacky to me, but wouldn't I get back my own modified shapes from `xcb_shape_query_extents` for all calls except for the initial one?

Maybe if I stored the shape pixmaps and compared them to the reply to see if `xcb_shape_query_extents` is giving me back my own shapes... Would this be very memory/process-heavy? I would be storing two pixmaps per window. I would also be comparing pixmaps many times a second when the user is resizing with their mouse.

**Comment by psychon** [comment]
`SHAPE` does not give you access to the pixmap, I think, but only to a list of rectangles describing the shape.
That's basically why I said that you would only do this check at the time you get a `MapRequest` event: If the window has a shape before it becomes visible, well, it is shaped. If it starts having a shape only later, then the race happens again.

Alternatively, but also uglily: You could track sequence numbers. Instead of `xcb_shape_set_shape_whatever(..)`, you would do `xcb_grab_server(); seq = set_shape().sequence; xcb_no_operation(); xcb_ungrab_server()`. You save the values `seq` in an array and when a `ShapeNotify` event comes in, you check if its sequence number is in the array or not to check if you or "something else" changed the shape.
However, this might still only detect the race after it already happened.

**Comment by Javyre** [comment]
Well I tried option 1 because it seems to me to be the simplest one to implement. But there are some issues:

- Multiple shape changes on the same window mapping seem to be squashed into a single `ShapeNotify` event per shape. This makes it hard/impossible to simply increase the count on `window_rounded_borders` and decrease it on `shape_notify`.

- The race condition is still present even if you would fix the above issue. The damage is already done once the wm stops reshaping the window.

So I guess either I'm missing something or this option should be discarded.

**Comment by psychon** [comment]
Sorry if you did not see my edit: https://github.com/baskerville/bspwm/issues/848#issuecomment-423175909

> Idea (1) only detects "uses shapes" after the shape was already corrupted.

...and then would e.g. "make things work" after a resize (where the WM now stops touching the shape).

**Comment by Javyre** [comment]
Alright. Time to move on to Idea (2).

But with this method I would have to wait for the first mapping of a window before starting applying shapes to it?

**Comment by re1** [comment]
> Basically, it wouldn't be possible without turning bspwm into a reparenting wm. And it would then require a compositor to be running.

I guess reparenting would also change the situation for some other issues, especially sided borders and title bars¹ as mentioned by @neeasade in #606.


¹ I am still unable to get colored title bars in bspwm

**Comment by Javyre** [comment]
@re1 Yes they would all be very possible. 

Though whether the features fit into bspwm's goals/needs/scope mainly depends on @baskerville I think.

**Comment by baskerville** [comment]
I only do checks at runtime. And I don't remember checking for any extensions besides *RandR* and *Xinerama*. Please note that *bspwm* already uses `xcb-shape` to make the preselection feedback transparent, this dependency was introduced by 3b96688.

**Comment by Javyre** [comment]
Wow I completely missed that. 
(I think my initial fork was out of date and I rebased it later not noticing the new xcb-shape dependency)
As you say, https://github.com/baskerville/bspwm/commit/3b96688d13ebc75e1a2883eda6efa74223359b37 doesn't do any checks so I guess I won't either.

**Comment by psychon** [comment]
> but I also want to know if anybody has some idea..

Shadows? Could it be that the window is larger than you think it is and the extra space is used to draw the window shadows?

**Comment by Javyre** [comment]
I killed my compositor and restarted firefox. Here is the result...
![](https://i.imgur.com/hvSDhoK.png)
As soon as i resize the window with my mouse (even by just one pixel) the bottom corners are properly rounded again. (this happens with & without my compositor running)

**Comment by baskerville** [comment]
The new preselection feedback should look like this:

![bspwm-preselection](https://user-images.githubusercontent.com/159419/46528844-10a91980-c895-11e8-8d2d-8c50f2647615.png)

It spans the entire preselected node, and is filled with two transparent colors that reflect the split ratio and direction. The most opaque region is where the new window will appear.

**Comment by baskerville** [comment]
> Are we assuming a compositor is always running alongside bspwm?

From time to time, users complain about the preselection feedback being a solid color: the problem is that it was designed with transparency in mind.

What might work is to have one window for each region (with different instance names).

Users of compton can then set the transparency of each region with opacity rules.

**Comment by Javyre** [comment]
That could work but then bspwm wouldn't fully work "right out of the box" anymore..

Maybe if we detect if a compositor is running and have the old style (half-opaque presel feedback window) as a fallback?

Then because we know at runtime if a compositor is running, we can have nice things like this with opacity and also anti-aliased round corners.

**Comment by baskerville** [comment]
> Maybe if we detect if a compositor is running and have the old style (half-opaque presel feedback window) as a fallback?

The p.f. should have two halves even when a compositor isn't running.

But never mind my request: I'll add this later, as an enhancement to the p.f.

The same remark goes for the ability to set irregular borders.

> Then because we know at runtime if a compositor is running, we can have nice things like this with opacity and also anti-aliased round corners.

Compositors might set properties on the root window.

Having a look at compton's code might be a good way to check that hypothesis.

**Comment by psychon** [comment]
Compositors aquire the `_NET_WM_SMs` selection to "announce their presence". I don't know if any specs actually require this, but see e.g.: https://github.com/awesomeWM/awesome/blob/c21e8b980a9e9b25fe9874c4f3fbb96f103bfc5e/luaa.c#L176-L211
(Disclaimer: Written by me)

**Comment by Javyre** [comment]
I think this is ready for a PR then now.

I couldn't find the cause of the firefox issue. I thought it was because the window starts as floating and doesn't get resized like other windows right away for tiling but it turns out it really only happens with firefox's download dialog...

@psychon yeah from what I've seen looking around compositors' code this seems to be at least an unofficial standard

@baskerville is reparenting + antialiased borders still on the table for a future PR?

**Comment by Javyre** [comment]
@sectioneight the issue is with the dialog before you choose to "save file"
Maybe this dialog didn't show up because you are saving an image on the page but you can try clicking `download zip` on the project main page and see what happens.

**Comment by Javyre** [comment]
Well at this point I'm basically just waiting on @baskerville for a response on the [PR](https://github.com/baskerville/bspwm/pull/856) and some of [the other topics](https://github.com/baskerville/bspwm/issues/848#issuecomment-427585379).

I'd be happy to work together with you for antialiased borders for example.

**Comment by sectioneight** [comment]
Sounds like a fun project if you're willing to take on a noob... the last time I did graphics programming was GLSL in college...

**Comment by baskerville** [comment]
This [post](https://venam.nixers.net/blog/unix/2018/05/10/reparenting-redirecting-composition-rendering.html), from the author of *2bwm*, might be helpful.

**Comment by nektie** [comment]
Is there any possibility for this PR to be merged any time soon?

**Comment by bandithijo** [comment]
some GTK and Qt application have this "firefox:Dialog" issue too when `state=floating` like "gnome-calculator", "spectacle". It can be emergency fix with add `rectangle=` rules. (^_^)

Another things I have found,
When I am using forking version from @Javyre on developer branch for rounded corner, everytime I click "fullscreen" button on web video player like YouTube, Vimeo, etc, it (web video player) did not fullscreen and make Firefox window state to fullscreen.

**Comment by exezin** [comment]
I also get the issue with firefox, can't seem to make any videos fullscreen and attempting to do so just bugs firefox out and requires me to restart it.  Works fine in chrome though.

Anyone aware of a workaround for the time being?

**Comment by lachlantula** [comment]
Any update on whether or not this is being merged sometime?

**Comment by BlueDrink9** [comment]
If this only has bugs in a couple of programs, and would be an option to turn on anyway, could it not be merged with warnings/caveats for those that use the affected programs?

**Comment by lachlantula** [comment]
https://github.com/baskerville/bspwm/pull/856#issuecomment-513130349

**Comment by BlueDrink9** [comment]
@clogic thoughts on why merging despite bugs is bad?

**Comment by MyNameIsCosmo** [comment]
The feature does not have to be merged to be used. There are comments throughout this thread and the PR that specify what needs to be fixed and tested. 

Feel free to fork Javyre's work, rebase to the current bspwm master, and work on fixing any issues. (Javyre's repository has issues disabled)

**Comment by ShaneEverittM** [comment]
RIP rounded corners?
